nixos/systemd: convert extraConfig to rfc 42

LordGrimmauld opened the pull request:
Converted `systemd.extraConfig`and `boot.initrd.systemd.extraConfig` to rfc-42-style `systemd.settings`.
Moved explicit definitions for `system.conf` to set settings instead.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and others READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

LordGrimmauld committed:
nixos/systemd: add settings.Manager option

LordGrimmauld committed:
nixos/testing: migrate to systemd.settings.Manager

LordGrimmauld committed:
nixos/tests/systemd: migrate to systemd.settings.Manager

LordGrimmauld committed:
nixos/pam: point to systemd.settings.Manager

LordGrimmauld committed:
nixos/tests/switchTest: migrate to systemd.settings.Manager

LordGrimmauld committed:
nixos/systemd: remove systemd.extraConfig

LordGrimmauld committed:
nixos/systemd: move systemd.managerEnvironment to systemd.settings.Manager.ManagerEnvironment

LordGrimmauld committed:
nixos/systemd: remove obsolete DefaultBlockIOAccounting option

This option is now being ignored by systemd,
so we do not need to explicitly define it.

LordGrimmauld committed:
nixos/systemd: make systemd.managerEnvironment affect systemd.settings.Manager

LordGrimmauld committed:
nixos/systemd: move systemd.watchdog.* to systemd.settings.Manager

LordGrimmauld committed:
nixos/systemd: move systemd.watchdog.* to systemd.settings.Manager

LordGrimmauld committed:
nixos/systemd: set DefaultLimitCORE in systemd.settings.Manager explicitly

LordGrimmauld committed:
nixos/systemd: remove obsolete definition for DefaultLimitCORE

The limit was introduced in 2016 in 840f3230a202e40a4107c4ce0ccb910be69c2792,
and broken iin 2019 in bafc25691596caf2f6e034cbe5933fdd90a92b6a. Since then,
it is the exact same as systemd itself sets as default.

LordGrimmauld committed:
nixos/systemd: explicitly set systemd.settings.Manager.Default*Accounting

LordGrimmauld committed:
nixos/systemd: remove obsolete DefaultCPUAccounting option

This option is made uncondiotional in systemd 258 [1].
Earlier, it defaulted to true on kernels newer than 4.15,
which applies to all supported nixos kernels.
This means removing the option does not change behavior.

[1] https://github.com/systemd/systemd/commit/29da53dde32a0097134f6c3ba5fb5220dd6e5423

LordGrimmauld committed:
nixos/systemd: remove enableCgroupAccounting option

ElvishJerricco reviewed (approved):
Code LGTM. We might want a release note, though I think you could argue the `mkRemovedOptionModule` and `mkRenamedOptionModule` take care of migration instructions sufficiently on their own.

@emilazy had mentioned on matrix that `systemd.settings` could potentially be an attrset of sections, but given that there's only one valid section I think that would just needlessly obscure the option documentation.

LordGrimmauld committed:
nixos/doc/rl-2511: document rfc42 conversion for systemd.extraConfig

emilazy commented on nixos/modules/system/boot/systemd.nix:
```diff
@@ -406,20 +407,16 @@ in
-      default = "";
-      type = types.lines;
-      example = "DefaultLimitCORE=infinity";
+    settings.Manager = mkOption {
```
I forget, is there a reason we can’t just set the defaults in `default` now? (Merging weirdness?)

This should probably have a `defaultText` showing the divergences we make from the stock configuration, in any case.

ElvishJerricco replied:
> I forget, is there a reason we can’t just set the defaults in `default` now? (Merging weirdness?)

With `default = { Foo = "bar"; };`, if any user sets `settings.Manager = /* literally anything */;` it will overwrite the defaults entirely and remove them.

emilazy replied:
Ah, yes. I forgot about that annoying detail.

emilazy commented on nixos/modules/system/boot/systemd.nix:
```diff
@@ -406,20 +407,16 @@ in
+      };
       description = ''
-        Extra config options for systemd. See {manpage}`systemd-system.conf(5)` man page
+        Options for the global systemd config. See {manpage}`systemd-system.conf(5)` man page
```
```suggestion
        Options for the global systemd service manager. See {manpage}`systemd-system.conf(5)` man page
```

emilazy commented on nixos/modules/system/boot/systemd.nix:
```diff
@@ -406,20 +407,16 @@ in
+      type = lib.types.submodule {
+        freeformType = types.attrsOf unitOption;
+      };
+      example = {
+        DefaultLimitCORE = "infinity";
+      };
```
Let’s not have a useless example. I’m fine just dropping it if we can’t think of something else to put in there. (`CtrlAltDelBurstAction = "none";`? The watchdog stuff?)

emilazy commented on nixos/modules/system/boot/systemd.nix:
```diff
@@ -750,6 +669,13 @@ in
       ) "${builtins.concatStringsSep ":" config.boot.extraSystemdUnitPaths}:";
     };
+    systemd.settings.Manager = {
+      ManagerEnvironment = lib.concatStringsSep " " (
+        lib.mapAttrsToList (n: v: "${n}=${lib.escapeShellArg v}") cfg.managerEnvironment
+      );
```
(I’d still like to see this RFC 42‐ized, but I agree that it can wait for another PR.)

emilazy commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -160,12 +168,16 @@ in
-      type = types.lines;
-      example = "DefaultLimitCORE=infinity";
+    settings.Manager = mkOption {
+      default = { };
+      type = lib.types.submodule {
+        freeformType = types.attrsOf unitOption;
+      };
+      example = {
+        DefaultLimitCORE = "infinity";
+      };
```
Same as above wrt. default and example.

emilazy commented on nixos/modules/system/boot/systemd/initrd.nix:
```diff
@@ -160,12 +168,16 @@ in
+      };
       description = ''
-        Extra config options for systemd. See {manpage}`systemd-system.conf(5)` man page
+        Options for the global systemd config used in initrd. See {manpage}`systemd-system.conf(5)` man page
```
```suggestion
        Options for the global systemd service manager used in initrd. See {manpage}`systemd-system.conf(5)` man page
```

emilazy commented on nixos/tests/switch-test.nix:
```diff
@@ -68,9 +68,9 @@ in
-          modifiedSystemConf.configuration.systemd.extraConfig = ''
-            # Hello world!
-          '';
+          modifiedSystemConf.configuration.systemd.settings.Manager = {
+            DummyTestConfig = "Hello World!";
+          };
```
This will start failing if we become able to test these configurations against the schema at build time.

```suggestion
          modifiedSystemConf.configuration.systemd.settings.Manager = {
            RuntimeWatchdogSec = "auto";
          };
```

or something? Or the other test does `DefaultEnvironment = "XXX_SYSTEM=foo";`.

emilazy reviewed (commented):
Thanks, this seems good to me; a few more nits and I’d be happy to see this landed.

`settings.Manager` feels weird to me but I can’t come up with any concrete problem it’ll cause to change it down the line. If/when we migrate `systemd.network.config` we’ll need a thing to produce the section structure, so we can adjust this then.

LordGrimmauld committed:
nixos/systemd: add boot.initrd.systemd.settings.Manager option

LordGrimmauld committed:
nixos/systemd: remove boot.initrd.systemd.extraConfig

LordGrimmauld committed:
nixos/systemd: make boot.initrd.systemd.managerEnvironment affect boot.initrd.systemd.settings.Manager

LordGrimmauld committed:
nixos/netdata: remove cgroup accounting enable

cgroup accounting is enabled by default, and the option is scheduled for removal.

emilazy reviewed (approved):
Looks great to me, thank you!

LordGrimmauld commented:
ideally both `networkd` config and `user.extraConfig` get converted, but there is currently a couple other things i am working on. I'll pick those up once i have time.
